### DISCUSSION TITLE
Error with refreshing probably due to shared expression

### DISCUSSION THREAD

**Discussion by Faber81** [TabularEditor/TabularEditor3#1415] [🙋♀️TE3 Q&A] [answered]
I created a new shared expression (SQL + M code that creates a list of text). The output of this expression is used to filter data in a query that generates the fact table. 

The code worked fine when I tried it in Power query but in TE3 I am unable to refresh the model. The error I get is very long and I cannot copy it.  

I tried to: 

-  Update table/model schema > I get an error 
![image](https://github.com/user-attachments/assets/e9ef4911-9623-4572-8fd0-d876316daa2f)

- Refresh the model/FACT table > Also get an error and the model does not refresh

Is there any action I am missing? It is the first time I create and use a shared expression.

The model works fine if change the code back to how it was (basically the different filtration logic in the FACT table query). 

Thanks!

**Comment by otykier** [comment]
From the error message it looks like the data source is not properly configured in the Power BI service. Could you sign in and check if perhaps credentials are missing?

E.g. in your Power BI workspace, click the ellipsis button next to the semantic model, then choose **Settings**, then locate the **Data source credentials** section and expand it. Let us know if you see any error messages there - and then try to click on **Edit credentials**:

<img width="511" alt="image" src="https://github.com/user-attachments/assets/94886dbe-3205-4a84-ae78-e6918c6877fb" />

**Reply by Faber81** [reply]
Hi Daniel, 
Thank you for your answer. I have taken over the semantic model (I was not the owner) but now I get a different error: 

![image](https://github.com/user-attachments/assets/6be4e5c1-142c-4c48-8379-b04425289def)


If I keep the old version of the code, it refreshes without problems. 
The list that is created by the shared expression is quite long and it goes all into a WHERE IN clause. 

btw, can you confirm I don't have to do any extra step when I create the shared expression? What I did was just writing the code and saving it.
Thank you!

**Reply by otykier** [reply]
Indeed, you shouldn't need to do anything other than add the shared expression, enter the M code, and save the model.

The screenshot, however, doesn't show the actual error message. Could you try to copy (Ctrl+C) all 3 lines in the Data Refresh view, and paste the content here? That way, we may be better able to assist in understanding the issue. Perhaps it's something as simple as invalid M syntax.

**Comment by Faber81** [comment, marked as answer]
Hi Daniel, 
It eventually worked, there was an issue with the credentials AND in the SQL code (a subtle one). It works now. 
Thanks for the time you spent helping me!
